### PR TITLE
feat(cypress): add cypress_web_test rule and @bazel/cypress package

### PR DESCRIPTION
**Pull request by mrmeku**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

Add @bazel/cypress for integrating cypress testing framework with bazel

Left todo:
Show how to launch a nodejs_binary server from cypress plugin

### CODE REVIEW

**Comment by mrmeku** [comment]
@alexeagle @gregmagolan time for a review! :D

File: examples/cypress/WORKSPACE [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+http_archive(
+    name = "build_bazel_rules_nodejs",
+    sha256 = "84abf7ac4234a70924628baa9a73a5a5cbad944c4358cf9abdb4aab29c9a5b77",
+    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/1.7.0/rules_nodejs-1.7.0.tar.gz"],
+)
+
+load("@build_bazel_rules_nodejs//:index.bzl", "yarn_install")
+
+yarn_install(
+    name = "npm",
```
Comment (alexeagle) [inline examples/cypress/WORKSPACE:31]: don't call it `npm` to verify that does'nt matter

File: packages/cypress/install.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Cypress rules for Bazel
+
+The Cypress rules run tests under the Cypress e2e testing framework with Bazel.
+
+## Installation
+
+Add the `@bazel/cypress` npm package to your `devDependencies` in `package.json`.
+Then, add `cypress verify` as a `postinstall` script in `package.json`.
```
Comment (alexeagle) [inline packages/cypress/install.md:8]: I might need to put something in `/WORKSPACE` too?

File: examples/cypress/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    testonly = True,
+    srcs = ["plugin.ts"],
+    tsconfig = ":tsconfig.json",
+    deps = [
+        "@npm//@types/node",
+        "@npm//express",
+    ],
+)
+
+ts_library(
```
Comment (alexeagle) [inline examples/cypress/BUILD.bazel:15]: any particular reason you chose ts_library rather than ts_project? I expect most users will prefer the latter since it's compatible with their existing code

File: examples/cypress/world.spec.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+describe('world', () => {
```
Comment (alexeagle) [inline examples/cypress/world.spec.js:1]: why is there a .ts and .js file both checked in? I suspect this one doesn't belong

File: package.json [hunk 5]
```diff
@@ -1,141 +1,145 @@
...
+    "@types/node": "^12.0.0",
+    "@types/semver": "6.2.0",
+    "babel-jest": "^25.5.1",
+    "bazel_workspaces": "file:./tools/npm_packages/bazel_workspaces",
+    "bazel_workspaces_consistent": "file:./tools/npm_packages/bazel_workspaces_consistent",
+    "c8": "7.1.0",
+    "clang-format": "1.2.2",
+    "conventional-changelog-cli": "^2.0.21",
+    "core-util-is": "^1.0.2",
+    "cypress": "^4.7.0",
```
Comment (alexeagle) [inline package.json:45]: I think you can revert this file? we don't require cypress anywhere in this workspace

File: packages/cypress/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    build_file_content = """exports_files([
+    "internal/cypress-install.js",
+    "internal/cypress_repository.bzl",
+    "internal/cypress_web_test.bzl",
+    "internal/plugins/base.js",
+    "internal/plugins/index.template.js",
+    "internal/run-cypress.js",
+])""",
+    substitutions = {
+        "@build_bazel_rules_nodejs//packages/cypress:internal/cypress-install.js": "@npm//@bazel/cypress:internal/cypress-install.js",
```
Comment (alexeagle) [inline packages/cypress/BUILD.bazel:83]: I suspect this can just be one entry
`//packages/cypress: //@bazel/cypress`

File: packages/cypress/internal/cypress-install.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
+const {spawnSync, spawn} = require('child_process');
```
Comment (alexeagle) [inline packages/cypress/internal/cypress-install.js:1]: add a comment about when does this code run? I guess it's part of the repository rule, since it writes .bzl files?

File: packages/typescript/test/reference_types_directive/BUILD.bazel [hunk 8]
```diff
@@ -18,8 +18,12 @@ ts_library(
     name = "test_lib",
     testonly = True,
     srcs = glob(["*.spec.ts"]),
-    node_modules = "@npm//:node_modules",
     tsconfig = ":tsconfig.json",
+    deps = [
```
Comment (alexeagle) [inline packages/typescript/test/reference_types_directive/BUILD.bazel:22]: unrelated change?

File: package.json (hunk 5 above)
Comment (alexeagle) [reply package.json:45]: oh I see it's used in the test. let's move cypress depnedency to packages/cypress/package.json and install that as cypress_deps, like packages/angular does?

File: .circleci/config.yml [hunk 9]
```diff
@@ -108,6 +108,7 @@ jobs:
           key: *cache_key
           paths:
             - "node_modules"
+            - "/home/circleci/.cache/Cypress"
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply .circleci/config.yml:111]: Actually, this can now be removed. My b!

File: WORKSPACE [hunk 10]
```diff
@@ -126,6 +126,13 @@ load("//packages/typescript/internal:ts_repositories.bzl", "ts_setup_workspace")
 
 ts_setup_workspace()
 
+#
+# Install @bazel/cypress dependencies
+#
+load("//packages/cypress:index.bzl", "cypress_repository")
+
+cypress_repository(name = "cypress")
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply WORKSPACE:134]: It is required for to run `bazel test //packages/cypress/...`

File: examples/cypress/.bazelrc [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import %workspace%/../../common.bazelrc
+
+build --experimental_inprocess_symlink_creation
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply examples/cypress/.bazelrc:3]: done

File: examples/cypress/BUILD.bazel (hunk 3 above)
Comment (mrmeku) [reply examples/cypress/BUILD.bazel:15]: The reason I opted for ts_library was to test that I consumed the ts_library providers correctly. Maybe I should use both a ts_library and ts_project? What do you think?

File: examples/cypress/world.spec.js (hunk 4 above)
Comment (mrmeku) [reply examples/cypress/world.spec.js:1]: I did so to show how to load a `.js` file as a test and as a `.ts` file as a test respectively

File: packages/cypress/internal/cypress-install.js (hunk 7 above)
Comment (mrmeku) [reply packages/cypress/internal/cypress-install.js:1]: done

File: examples/cypress/WORKSPACE (hunk 1 above)
Comment (mrmeku) [reply examples/cypress/WORKSPACE]: done

File: packages/cypress/install.md (hunk 2 above)
Comment (mrmeku) [reply packages/cypress/install.md]: done

File: .bazelrc [hunk 12]
```diff
@@ -17,6 +17,7 @@ build:release --stamp --workspace_status_command=./scripts/current_version.sh
 build --define=SOME_TEST_ENV=some_value
 test --define=SOME_TEST_ENV=some_value
 build --action_env=SOME_OTHER_ENV=some_other_value
+build --experimental_inprocess_symlink_creation
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply .bazelrc:20]: done (in the readme.md)

File: package.json (hunk 5 above)
Comment (mrmeku) [reply package.json]: done

File: packages/typescript/test/reference_types_directive/BUILD.bazel (hunk 8 above)
Comment (mrmeku) [reply packages/typescript/test/reference_types_directive/BUILD.bazel]: removed

File: packages/cypress/BUILD.bazel (hunk 6 above)
Comment (mrmeku) [reply packages/cypress/BUILD.bazel]: I ended up needing to be specific since I have to do a different workspace for `cypress_repository.bzl`

File: .bazelci/presubmit.yml [hunk 13]
```diff
@@ -120,6 +122,8 @@ tasks:
     - "@bazel_workspace_b//subdir:bin"
     build_targets:
     - "//..."
+    build_flags:
+    - "--build_tag_filters=-cypress"
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply .bazelci/presubmit.yml:130]: commented in the file

File: WORKSPACE (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrmeku) [reply WORKSPACE]: I think if you were to remove it and then do a bazel clean the test would then fail

File: examples/cypress/BUILD.bazel [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    ],
+)
+
+cypress_web_test(
+    name = "test",
+    srcs = [
+        "world.spec.js",
+        ":hello_spec",
+    ],
+    config_file = "cypress.json",
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply examples/cypress/BUILD.bazel:37]: done

File: examples/cypress/plugin.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import * as express from 'express';
+
+
+module.exports = (on, config) => {
```
[the remark this answers is not in the document]
Comment (mrmeku) [reply examples/cypress/plugin.ts:5]: done

File: examples/cypress/world.spec.js (hunk 4 above)
Comment (alexeagle) [reply examples/cypress/world.spec.js:1]: ah, okay then I would want them to have different filenames so it's clear one isn't derived from the other

File: examples/cypress/BUILD.bazel (hunk 3 above)
Comment (alexeagle) [reply examples/cypress/BUILD.bazel:17]: we shouldn't have a difference between providers, such that you can tell whether ts_project or ts_library was used to compile your dependency. Could you file an issue with the details about what's different? I hope you don't still need the "typescript" provider
